File: Data Structures & Algorithms/balanced-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        int res = findHeight(root);
        if (res == -1) {
            return false;
        }
        return true;
    }

    int findHeight(TreeNode* root) {
        if (!root) {
            return 0;
        }

        int leftHeight = findHeight(root->left);
        int rightHeight = findHeight(root->right);

        if (leftHeight == -1 || rightHeight == -1) {
            return -1;
        }

        if (std::abs(rightHeight - leftHeight) > 1) {
            return -1;
        }

        return 1 + max(leftHeight, rightHeight);



    }
};

File: Data Structures & Algorithms/binary-tree-diameter/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int: 
        self.res = 0     
        def findPath(root):
            if not root:
                return 0

            leftHeight = findPath(root.left)
            rightHeight = findPath(root.right)

            self.res = max(self.res, leftHeight + rightHeight)
            return 1 + max(leftHeight, rightHeight)

        findPath(root)
        return self.res